File: examples/php/UserManager.php
<?php

/**
 * 用户管理类
 */
namespace TestMind;

class User {
    public $id;
    public $username;
    public $email;
    public $password;
    public $firstName;
    public $lastName;
    public $active;

    public function __construct($username, $email, $password, $firstName = null, $lastName = null, $active = true) {
        $this->id = null;
        $this->username = $username;
        $this->email = $email;
        $this->password = $password;
        $this->firstName = $firstName;
        $this->lastName = $lastName;
        $this->active = $active;
    }

    public function getFullName() {
        if ($this->firstName && $this->lastName) {
            return $this->firstName . ' ' . $this->lastName;
        } elseif ($this->firstName) {
            return $this->firstName;
        } elseif ($this->lastName) {
            return $this->lastName;
        }
        return $this->username;
    }
}

class UserManager {
    private $users = [];
    private $nextId = 1;

    public function findUserById($id) {
        return isset($this->users[$id]) ? $this->users[$id] : null;
    }

    public function findAllUsers() {
        return array_values($this->users);
    }

    public function createUser($user) {
        if (empty($user->email)) {
            return null;
        }

        if (!$this->validateEmail($user->email)) {
            return null;
        }

        $user->id = $this->nextId++;
        $this->users[$user->id] = $user;
        return $user;
    }

    public function updateUser($id, $user) {
        if ($id <= 0 || !isset($this->users[$id])) {
            return null;
        }

        $user->id = $id;
        $this->users[$id] = $user;
        return $user;
    }

    public function deleteUser($id) {
        if ($id <= 0 || !isset($this->users[$id])) {
            return false;
        }

        unset($this->users[$id]);
        return true;
    }

    public function activateUser($id) {
        $user = $this->findUserById($id);
        if (!$user) {
            return false;
        }

        $user->active = true;
        return true;
    }

    public function deactivateUser($id) {
        $user = $this->findUserById($id);
        if (!$user) {
            return false;
        }

        $user->active = false;
        return true;
    }

    public function findUserByEmail($email) {
        foreach ($this->users as $user) {
            if ($user->email === $email) {
                return $user;
            }
        }
        return null;
    }

    public function findUserByUsername($username) {
        foreach ($this->users as $user) {
            if ($user->username === $username) {
                return $user;
            }
        }
        return null;
    }

    public function searchUsers($keyword, $options = []) {
        $searchUsername = !isset($options['username']) || $options['username'];
        $searchEmail = !isset($options['email']) || $options['email'];
        $searchName = !isset($options['name']) || $options['name'];

        $results = [];
        $lowerKeyword = strtolower($keyword);

        foreach ($this->users as $user) {
            $match = false;

            if ($searchUsername) {
                $match = strpos(strtolower($user->username), $lowerKeyword) !== false;
            }

            if (!$match && $searchEmail) {
                $match = strpos(strtolower($user->email), $lowerKeyword) !== false;
            }

            if (!$match && $searchName) {
                if ($user->firstName && strpos(strtolower($user->firstName), $lowerKeyword) !== false) {
                    $match = true;
                } elseif ($user->lastName && strpos(strtolower($user->lastName), $lowerKeyword) !== false) {
                    $match = true;
                }
            }

            if ($match) {
                $results[] = $user;
            }
        }

        return $results;
    }

    public function createUsersBatch($users) {
        $created = [];
        foreach ($users as $user) {
            $createdUser = $this->createUser($user);
            if ($createdUser) {
                $created[] = $createdUser;
            }
        }
        return $created;
    }

    public function deleteUsersBatch($ids) {
        $deletedCount = 0;
        foreach ($ids as $id) {
            if ($this->deleteUser($id)) {
                $deletedCount++;
            }
        }
        return $deletedCount;
    }

    public function getActiveUserCount() {
        $count = 0;
        foreach ($this->users as $user) {
            if ($user->active) {
                $count++;
            }
        }
        return $count;
    }

    public function getTotalUserCount() {
        return count($this->users);
    }

    private function validateEmail($email) {
        return filter_var($email, FILTER_VALIDATE_EMAIL) !== false;
    }
}

?>

File: examples/ruby/user_manager.rb
# 用户管理模块
# 提供用户管理功能
module TestMind
  class User
    attr_accessor :id, :username, :email, :password, :first_name, :last_name, :active

    def initialize(username, email, password, first_name = nil, last_name = nil, active = true)
      @id = nil
      @username = username
      @email = email
      @password = password
      @first_name = first_name
      @last_name = last_name
      @active = active
    end

    def full_name
      if @first_name && @last_name
        "#{@first_name} #{@last_name}"
      elsif @first_name
        @first_name
      elsif @last_name
        @last_name
      else
        @username
      end
    end
  end

  class UserManager
    def initialize
      @users = {}
      @next_id = 1
      @mutex = Mutex.new
    end

    # 根据ID查找用户
    def find_user_by_id(id)
      @mutex.synchronize do
        @users[id]
      end
    end

    # 获取所有用户
    def find_all_users
      @mutex.synchronize do
        @users.values
      end
    end

    # 创建新用户
    def create_user(user)
      return nil if user.email.nil? || user.email.empty?
      return nil unless valid_email?(user.email)

      @mutex.synchronize do
        user.id = @next_id
        @next_id += 1
        @users[user.id] = user
        user
      end
    end

    # 更新用户信息
    def update_user(id, user)
      return nil if id <= 0

      @mutex.synchronize do
        return nil unless @users.key?(id)

        user.id = id
        @users[id] = user
        user
      end
    end

    # 删除用户
    def delete_user(id)
      return false if id <= 0

      @mutex.synchronize do
        !!@users.delete(id)
      end
    end

    # 激活用户
    def activate_user(id)
      user = find_user_by_id(id)
      return false unless user

      user.active = true
      true
    end

    # 停用用户
    def deactivate_user(id)
      user = find_user_by_id(id)
      return false unless user

      user.active = false
      true
    end

    # 根据邮箱查找用户
    def find_user_by_email(email)
      @mutex.synchronize do
        @users.values.find { |user| user.email == email }
      end
    end

    # 根据用户名查找用户
    def find_user_by_username(username)
      @mutex.synchronize do
        @users.values.find { |user| user.username == username }
      end
    end

    # 搜索用户
    def search_users(keyword, options = {})
      search_username = options[:username] != false
      search_email = options[:email] != false
      search_name = options[:name] != false

      results = []
      lower_keyword = keyword.downcase

      @mutex.synchronize do
        @users.values.each do |user|
          match = false

          if search_username
            match = true if user.username.downcase.include?(lower_keyword)
          end

          if !match && search_email
            match = true if user.email.downcase.include?(lower_keyword)
          end

          if !match && search_name
            if user.first_name && user.first_name.downcase.include?(lower_keyword)
              match = true
            elsif user.last_name && user.last_name.downcase.include?(lower_keyword)
              match = true
            end
          end

          results << user if match
        end
      end

      results
    end

    # 批量创建用户
    def create_users_batch(users)
      created = []
      users.each do |user|
        created_user = create_user(user)
        created << created_user if created_user
      end
      created
    end

    # 批量删除用户
    def delete_users_batch(ids)
      deleted_count = 0
      ids.each do |id|
        deleted_count += 1 if delete_user(id)
      end
      deleted_count
    end

    # 获取活跃用户数量
    def active_user_count
      @mutex.synchronize do
        @users.values.count { |user| user.active }
      end
    end

    # 获取用户总数
    def total_user_count
      @mutex.synchronize do
        @users.size
      end
    end

    private

    def valid_email?(email)
      email =~ /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
    end
  end
end
